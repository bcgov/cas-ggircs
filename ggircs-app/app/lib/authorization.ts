import pagesAuthorization from "../data/pagesAuthorization.json";
import { match } from "path-to-regexp";

export const isRouteAuthorized = (
  route: string,
  userGroups: readonly string[]
) => {
  const authRules = pagesAuthorization.find(({ routePaths }) =>
    routePaths.some((routePath) =>
      match(routePath, { decode: decodeURIComponent, endsWith: "?" })(route)
    )
  );

  if (!authRules) {
    return false;
  }

  const { isProtected, allowedRoles = [] } = authRules;

  if (!isProtected) {
    return true;
  }

  return allowedRoles.some((role) => userGroups.includes(role));
};
