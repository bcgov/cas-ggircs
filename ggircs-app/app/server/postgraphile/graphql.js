const {
  createPostGraphileSchema,
  withPostGraphileContext,
} = require("postgraphile");
const { graphql } = require("graphql");
const authenticationPgSettings = require("./authenticationPgSettings");
const { dbPool, getDatabaseUrl } = require("../storage/db");
const postgraphileOptions = require("./postgraphileOptions");

const pgSettings = (req) => {
  const opts = {
    ...authenticationPgSettings(req),
  };
  return opts;
};

let postgraphileSchemaSingleton;

const postgraphileSchema = async () => {
  if (!postgraphileSchemaSingleton) {
    postgraphileSchemaSingleton = await createPostGraphileSchema(
      getDatabaseUrl(),
      process.env.DATABASE_SCHEMA || "ggircs_app",
      postgraphileOptions()
    );
  }

  return postgraphileSchemaSingleton;
};

async function performQuery(query, variables, request) {
  const settings = pgSettings(request);
  return withPostGraphileContext(
    {
      dbPool,
      pgSettings: settings,
    },
    async (context) => {
      // Execute your GraphQL query in this function with the provided
      // `context` object, which should NOT be used outside of this
      // function.
      const result = graphql(
        await postgraphileSchema(),
        query,
        null,
        { ...context }, // You can add more to context if you like
        variables
      );
      return result;
    }
  );
}

module.exports = {
  performQuery,
};
