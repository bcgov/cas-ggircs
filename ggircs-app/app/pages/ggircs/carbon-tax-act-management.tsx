import { withRelay, RelayProps } from "relay-nextjs";
import { graphql, usePreloadedQuery } from "react-relay/hooks";
import withRelayOptions from "lib/relay/withRelayOptions";
import type { NextPageContext } from "next";
import { NextRouter } from "next/router";
import { carbonTaxActManagementQuery } from "__generated__/carbonTaxActManagementQuery.graphql";
import DefaultLayout from "components/Layout/DefaultLayout";
import CarbonTaxActFuelType from "components/FuelManagement/CarbonTaxActFuelType";

export const CarbonTaxActManagementQuery = graphql`
  query carbonTaxActManagementQuery($carbonTaxActFuelTypeId: ID!) {
    query {
      ...CarbonTaxActFuelType_query
        @arguments(carbonTaxActFuelTypeId: $carbonTaxActFuelTypeId)
      session {
        ...DefaultLayout_session
      }
    }
  }
`;

export function CarbonTaxActManagement({
  preloadedQuery,
}: RelayProps<{}, carbonTaxActManagementQuery>) {
  const { query } = usePreloadedQuery(
    CarbonTaxActManagementQuery,
    preloadedQuery
  );

  return (
    <>
      <DefaultLayout
        session={query.session}
        title="Fuel Type Management"
        width="wide"
      >
        <div>
          <CarbonTaxActFuelType
            query={query}
            pageQuery={CarbonTaxActManagementQuery}
          />
        </div>
      </DefaultLayout>
    </>
  );
}

const withRelayOptionsOverride = {
  ...withRelayOptions,
  variablesFromContext: (ctx: NextPageContext | NextRouter) => ({
    carbonTaxActFuelTypeId: "",
    showAll: "false",
    ...ctx.query,
  }),
};

export default withRelay(
  CarbonTaxActManagement,
  CarbonTaxActManagementQuery,
  withRelayOptionsOverride
);
