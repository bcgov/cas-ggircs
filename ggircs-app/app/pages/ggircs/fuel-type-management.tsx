import { withRelay, RelayProps } from "relay-nextjs";
import { graphql, usePreloadedQuery } from "react-relay/hooks";
import withRelayOptions from "lib/relay/withRelayOptions";
import type { NextPageContext } from "next";
import { NextRouter } from "next/router";
import { fuelTypeManagementQuery } from "__generated__/fuelTypeManagementQuery.graphql";
import DefaultLayout from "components/Layout/DefaultLayout";
import NormalizedFuelType from "components/FuelManagement/NormalizedFuelType";
import UnmappedFuelTypes from "components/FuelManagement/UnmappedFuelTypes";

export const FuelTypeManagementQuery = graphql`
  query fuelTypeManagementQuery($fuelCarbonTaxDetailId: ID!) {
    query {
      ...NormalizedFuelType_query
        @arguments(fuelCarbonTaxDetailId: $fuelCarbonTaxDetailId)
      ...UnmappedFuelTypes_query
      session {
        ...DefaultLayout_session
      }
    }
  }
`;

export function FuelTypeManagement({
  preloadedQuery,
}: RelayProps<{}, fuelTypeManagementQuery>) {
  const { query } = usePreloadedQuery(FuelTypeManagementQuery, preloadedQuery);
  return (
    <>
      <DefaultLayout
        session={query.session}
        title="Fuel Type Management"
        width="wide"
      >
        <div id="unmapped">
          <UnmappedFuelTypes query={query} />
        </div>
        <div>
          <NormalizedFuelType query={query} />
        </div>
      </DefaultLayout>
      <style jsx>
        {`
          #unmapped {
            margin-bottom: 2em;
          }
        `}
      </style>
    </>
  );
}

const withRelayOptionsOverride = {
  ...withRelayOptions,
  variablesFromContext: (ctx: NextPageContext | NextRouter) => ({
    fuelCarbonTaxDetailId: "",
    ...ctx.query,
  }),
};

export default withRelay(
  FuelTypeManagement,
  FuelTypeManagementQuery,
  withRelayOptionsOverride
);
