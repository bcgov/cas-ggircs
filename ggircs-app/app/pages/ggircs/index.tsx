import DefaultLayout from "components/Layout/DefaultLayout";
import { withRelay, RelayProps } from "relay-nextjs";
import { graphql, usePreloadedQuery } from "react-relay/hooks";
import { ggircsLandingQuery } from "__generated__/ggircsLandingQuery.graphql";
import withRelayOptions from "lib/relay/withRelayOptions";
import SwrsDataAccessCard from "components/Dashboard/SwrsDataAccessCard";

export const GgircsLandingQuery = graphql`
  query ggircsLandingQuery {
    query {
      session {
        ...DefaultLayout_session
      }
    }
  }
`;

export function GgircsLanding({ preloadedQuery }: RelayProps<{}, ggircsLandingQuery>) {
  const { query } = usePreloadedQuery(GgircsLandingQuery, preloadedQuery);
  return (
    <DefaultLayout session={query.session}>
      <SwrsDataAccessCard />
    </DefaultLayout>
  );
}

export default withRelay(GgircsLanding, GgircsLandingQuery, withRelayOptions);
