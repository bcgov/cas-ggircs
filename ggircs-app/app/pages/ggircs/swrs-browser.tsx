import { withRelay, RelayProps } from "relay-nextjs";
import { graphql, usePreloadedQuery } from "react-relay/hooks";
import withRelayOptions from "lib/relay/withRelayOptions";
import { swrsBrowserQuery } from "__generated__/swrsBrowserQuery.graphql";
import DefaultLayout from "components/Layout/DefaultLayout";
import FileList from "components/SwrsBrowser/FileList";

export const SwrsBrowserQuery = graphql`
  query swrsBrowserQuery {
    query {
      session {
        ...DefaultLayout_session
      }
    }
  }
`;

export function SwrsBrowser({ preloadedQuery }: RelayProps<{}, swrsBrowserQuery>) {
  const { query } = usePreloadedQuery(SwrsBrowserQuery, preloadedQuery);
  return (
    <DefaultLayout
      session={query.session}
      title="ECCC SWRS File Explorer"
      width="wide"
    >
      <FileList />
    </DefaultLayout>
  );
}

export default withRelay(SwrsBrowser, SwrsBrowserQuery, withRelayOptions);
