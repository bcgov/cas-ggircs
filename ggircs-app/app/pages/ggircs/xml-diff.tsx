import { withRelay, RelayProps } from "relay-nextjs";
import { graphql, usePreloadedQuery } from "react-relay/hooks";
import { Row, Col } from "react-bootstrap";
import withRelayOptions from "lib/relay/withRelayOptions";
import { xmlDiffQuery } from "__generated__/xmlDiffQuery.graphql";
import DefaultLayout from "components/Layout/DefaultLayout";
import ReportSelector from "components/XmlDiff/ReportSelector";
import DiffDetailsContainer from "components/XmlDiff/DiffDetailsContainer";
import RenderDiff from "components/XmlDiff/RenderDiff";
import DiffControls from "components/XmlDiff/DiffControls";
import { useRouter } from "next/router";
import type { NextPageContext } from "next";
import { NextRouter } from "next/router";

export const XmlDiffQuery = graphql`
  query xmlDiffQuery($FirstSideRelayId: ID!, $SecondSideRelayId: ID!) {
    query {
      session {
        ...DefaultLayout_session
      }
      allReports {
        edges {
          node {
            id
            swrsReportId
          }
        }
      }
      ...DiffDetailsContainer_query
        @arguments(
          FirstSideRelayId: $FirstSideRelayId
          SecondSideRelayId: $SecondSideRelayId
        )
      ...RenderDiff_query
        @arguments(
          FirstSideRelayId: $FirstSideRelayId
          SecondSideRelayId: $SecondSideRelayId
        )
    }
  }
`;

export function XmlDiff({ preloadedQuery }: RelayProps<{}, xmlDiffQuery>) {
  const { query } = usePreloadedQuery(XmlDiffQuery, preloadedQuery);

  // We need to ensure both ReportSelector components share the same router, otherwise they
  // can't write on the same query in the URL.
  const router = useRouter();

  return (
    <DefaultLayout
      session={query.session}
      title="ECCC SWRS XML Diff"
      width="wide"
    >
      <Row>
        <Col md={{ span: 6, order: router.query.reversed ? 2 : 1 }}>
          <ReportSelector
            diffSide={router.query.reversed ? "Second" : "First"}
            allReports={query.allReports.edges}
            router={router}
          />
        </Col>
        <Col md={{ span: 6, order: router.query.reversed ? 1 : 2 }}>
          <ReportSelector
            diffSide={router.query.reversed ? "First" : "Second"}
            allReports={query.allReports.edges}
            router={router}
          />
        </Col>
      </Row>
      <DiffControls />

      <DiffDetailsContainer query={query} />

      <RenderDiff query={query} />
    </DefaultLayout>
  );
}

const withRelayOptionsOverride = {
  ...withRelayOptions,
  variablesFromContext: (ctx: NextPageContext | NextRouter) => ({
    FirstSideRelayId: "",
    SecondSideRelayId: "",
    ...ctx.query,
  }),
};

export default withRelay(XmlDiff, XmlDiffQuery, withRelayOptionsOverride);
