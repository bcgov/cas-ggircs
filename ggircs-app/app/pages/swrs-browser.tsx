import React, { Component } from "react";
import { graphql } from "react-relay";
import { swrsBrowserQueryResponse } from "swrsBrowserQuery.graphql";
import { PageComponentProps } from "next-env";
import DefaultLayout from "components/Layout/DefaultLayout";
import FileList from "components/SwrsBrowser/FileList";
import { USER_GROUP } from "data/group-constants";

const ALLOWED_GROUPS = [...USER_GROUP];

interface Props extends PageComponentProps {
  query: swrsBrowserQueryResponse["query"];
}
export default class Index extends Component<Props> {
  static allowedGroups = ALLOWED_GROUPS;
  static isAccessProtected = true;
  static query = graphql`
    query swrsBrowserQuery {
      query {
        session {
          ...DefaultLayout_session
        }
      }
    }
  `;

  render() {
    const { query } = this.props;
    const { session } = query || {};

    return (
      <DefaultLayout session={session} title="ECCC SWRS File Explorer">
        <FileList />
      </DefaultLayout>
    );
  }
}
