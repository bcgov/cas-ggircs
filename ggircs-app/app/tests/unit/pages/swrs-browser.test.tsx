import React from "react";
import { SwrsBrowser, SwrsBrowserQuery } from "../../../pages/ggircs/swrs-browser";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import {
  createMockEnvironment,
  MockPayloadGenerator,
  RelayMockEnvironment,
} from "relay-test-utils";
import { RelayEnvironmentProvider, loadQuery } from "react-relay";
import { swrsBrowserQuery } from "__generated__/swrsBrowserQuery.graphql";
import { MockResolvers } from "relay-test-utils/lib/RelayMockPayloadGenerator";

let environment: RelayMockEnvironment;
let initialQueryRef;

const defaultMockResolver = {
  Query() {
    return {
      session: { ggircsUserBySub: {} },
    };
  },
};

const loadTestQuery = (mockResolver: MockResolvers = defaultMockResolver) => {
  environment.mock.queueOperationResolver((operation) => {
    return MockPayloadGenerator.generate(operation, mockResolver);
  });

  environment.mock.queuePendingOperation(SwrsBrowserQuery, {});
  initialQueryRef = loadQuery<swrsBrowserQuery>(
    environment,
    SwrsBrowserQuery,
    {}
  );
};

const renderComponentUnderTest = () =>
  render(
    <RelayEnvironmentProvider environment={environment}>
      <SwrsBrowser CSN preloadedQuery={initialQueryRef} />
    </RelayEnvironmentProvider>
  );

describe("The swrs-browser page", () => {
  beforeEach(() => {
    environment = createMockEnvironment();
  });

  it("renders the swrs-browser page", () => {
    loadTestQuery();
    renderComponentUnderTest();

    expect(
      screen.getByText(
        /ECCC SWRS File Explorer/i
      )
    ).toBeInTheDocument();
  });
});

