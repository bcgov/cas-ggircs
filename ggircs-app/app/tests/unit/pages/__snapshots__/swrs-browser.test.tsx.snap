// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The swrs-browser page matches the last accepted Snapshot 1`] = `
<Relay(DefaultLayout)
  session={null}
  title="ECCC SWRS File Explorer"
>
  <FileList />
</Relay(DefaultLayout)>
`;
